pkg/handlers/generic/mutation/taints/inject_controlplane.go

```diff
@@ -0,0 +1,107 @@
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/v1alpha1"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/handlers/mutation"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches/selectors"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/variables"
+)
+
+type taintsControlPlanePatchHandler struct {
+	variableName      string
+	variableFieldPath []string
+}
+
+func NewControlPlanePatch() *taintsControlPlanePatchHandler {
+	return newTaintsControlPlanePatchHandler(
+		v1alpha1.ClusterConfigVariableName,
+		v1alpha1.ControlPlaneConfigVariableName,
+		VariableName,
+	)
+}
+
+func newTaintsControlPlanePatchHandler(
+	variableName string,
+	variableFieldPath ...string,
+) *taintsControlPlanePatchHandler {
+	return &taintsControlPlanePatchHandler{
+		variableName:      variableName,
+		variableFieldPath: variableFieldPath,
+	}
+}
+
+func (h *taintsControlPlanePatchHandler) Mutate(
+	ctx context.Context,
+	obj *unstructured.Unstructured,
+	vars map[string]apiextensionsv1.JSON,
+	holderRef runtimehooksv1.HolderReference,
+	_ ctrlclient.ObjectKey,
+	_ mutation.ClusterGetter,
+) error {
+	log := ctrl.LoggerFrom(ctx).WithValues(
+		"holderRef", holderRef,
+	)
+
+	taintsVar, err := variables.Get[[]v1alpha1.Taint](
+		vars,
+		h.variableName,
+		h.variableFieldPath...,
+	)
+	if err != nil {
+		if variables.IsNotFoundError(err) {
+			log.V(5).Info("Taints variable for worker not defined")
+			return nil
+		}
+		return err
+	}
+
+	log = log.WithValues(
+		"variableName",
+		h.variableName,
+		"variableFieldPath",
+		h.variableFieldPath,
+		"variableValue",
+		taintsVar,
+	)
+
+	return patches.MutateIfApplicable(
+		obj, vars, &holderRef, selectors.ControlPlane(), log,
+		func(obj *controlplanev1.KubeadmControlPlaneTemplate) error {
+			log.WithValues(
+				"patchedObjectKind", obj.GetObjectKind().GroupVersionKind().String(),
+				"patchedObjectName", ctrlclient.ObjectKeyFromObject(obj),
+			).Info("adding taints to worker node kubeadm config template")
+			if obj.Spec.Template.Spec.KubeadmConfigSpec.InitConfiguration != nil {
+				obj.Spec.Template.Spec.KubeadmConfigSpec.InitConfiguration = &bootstrapv1.InitConfiguration{}
+			}
+			if obj.Spec.Template.Spec.KubeadmConfigSpec.JoinConfiguration != nil {
+				obj.Spec.Template.Spec.KubeadmConfigSpec.JoinConfiguration = &bootstrapv1.JoinConfiguration{}
+			}
+			coreTaints := toCoreTaints(taintsVar)
+
+			obj.Spec.Template.Spec.KubeadmConfigSpec.InitConfiguration.NodeRegistration.Taints = append(
```

Comment:
>taints specifies the taints the Node API object should be registered with. If this field is unset, i.e. nil, it will be defaulted with a control-plane taint for control-plane nodes. 

From the docs https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta4/#kubeadm-k8s-io-v1beta4-NodeRegistrationOptions 

If we do set a taint here, do the default control-plane taints still get added?
Also do we want to support the usecase of setting an empty taints and letting users remove the defaults?

Reply:
You're suspicion is correct: default CP taints do not get added...

```
<SNIP>
labels:
      node-role.kubernetes.io/control-plane: ""
<SNIP>
spec:
    taints:
    - effect: NoSchedule
      key: some-key
      value: some-value
```

Is that what we want here? I suggest tainting the CP nodes is less common than for worker nodes so I think this is OK, and indeed doing what you suggest if the taints are set but empty slice would allow removing the default taints, allowing scheduling in constrained resource environments.

Reply:
>To support this would require a larger change that I would not be happy doing in this PR at least.

That makes sense, users can still do this by setting it to an empty slice in the KCP Template.

Reply:
> > To support this would require a larger change that I would not be happy doing in this PR at least.
> 
> That makes sense, users can still do this by setting it to an empty slice in the KCP Template.

Or by adding an inline JSON patch to the ClusterClass.

Reply:
Added docs.

Reply:
FIgured it out! We now support not setting default taints! Please take another look @dkoshkin, also added docs about it.